File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.py
import sys

sys.setrecursionlimit(10 ** 5)

while True:
    n, *arr = map(int, input().split())

    if n == 0:
        break

    stack = []
    result = 0

    for i in range(n):
        while stack and arr[i] < arr[stack[-1]]:
            h = arr[stack.pop()]
            w = i if not stack else i - stack[-1] - 1
            result = max(result, h * w)

        stack.append(i)

    while stack:
        h = arr[stack.pop()]
        w = n if not stack else n - stack[-1] - 1
        result = max(result, h * w)

    print(result)

File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.java
import java.io.*;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        Stack<Integer> stack = new Stack<>();

        while (true) {
            st = new StringTokenizer(br.readLine());
            int n = Integer.parseInt(st.nextToken());
            if (n == 0) break;
            int[] arr = new int[n];

            for (int i = 0; i < n; i++) {
                arr[i] = Integer.parseInt(st.nextToken());
            }

            long max = 0;
            for (int i = 0; i < n; i++) {
                while (!stack.isEmpty() && arr[stack.peek()] > arr[i]) {
                    int height = arr[stack.pop()];
                    int width = i;
                    if (!stack.isEmpty()) {
                        width = (i - stack.peek() - 1);
                    }
                    max = Math.max(max, (long) height * width);
                }
                stack.push(i);
            }
            while (!stack.isEmpty()) {
                int height = arr[stack.pop()];
                int width = n;
                if (!stack.isEmpty()) {
                    width = (n - stack.peek() - 1);
                }
                max = Math.max(max, (long) height * width);
            }
            bw.write(max + "\n");
        }
        bw.close();
    }
}